src/boot/fwupd.c

```diff
@@ -0,0 +1,567 @@
+                if (err != EFI_SUCCESS) {
+                        log_error("Failed to find capsule '%ls': %m", (*u)->name);
+                        continue;
+                }
+
+                _cleanup_file_close_ EFI_FILE *root = NULL;
+                err = open_volume(device_handle, &root);
+                if (err != EFI_SUCCESS) {
+                        log_error("Failed to open volume for '%ls': %m", (*u)->name);
+                        continue;
+                }
+
+                _cleanup_free_ char16_t *dp_str = NULL;
+                err = device_path_to_str(file_dp, &dp_str);
+                if (err != EFI_SUCCESS) {
+                        log_error("Failed to convert device path for '%ls' to string: %m", (*u)->name);
+                        continue;
+                }
+
+                _cleanup_free_ char *file_buffer = NULL;
+                size_t file_size = 0;
+                err = file_read(root, dp_str, 0, 0, &file_buffer, &file_size);
+                if (err != EFI_SUCCESS) {
+                        log_error("Failed to read file for '%ls': %m", (*u)->name);
+                        continue;
+                }
+
+                if (file_size < sizeof(EFI_CAPSULE_HEADER)) {
+                        log_error("File '%ls' is too small for EFI_CAPSULE_HEADER: %zu < %zu", (*u)->name, file_size, sizeof(EFI_CAPSULE_HEADER));
+                        continue;
+                }
+
+                EFI_CAPSULE_BLOCK_DESCRIPTOR block_descriptor = {
+                        .Length = file_size,
+                        .DataBlock = (EFI_PHYSICAL_ADDRESS)(uintptr_t)file_buffer,
+                };
+                EFI_CAPSULE_HEADER *capsule = (EFI_CAPSULE_HEADER *)file_buffer;
+
+                EFI_GUID guid = (*u)->info->guid; /* Avoid unaligned access warning */
+                if (efi_guid_equal(&guid, MAKE_GUID_PTR(UX_CAPSULE))) {
+                        UX_CAPSULE_HEADER *hdr = (UX_CAPSULE_HEADER *)(((uint8_t *)capsule) + capsule->HeaderSize);
+                        uint32_t mode = 0;
+
+                        err = get_graphics_protocol_mode(image, &mode);
+                        if (err != EFI_SUCCESS)
+                                return log_error_status(err, "Failed to get graphics protocol mode for UX capsule: %m");
+                        hdr->mode = mode; /* Avoid unaligned access warning */
+
+                        /* Since the mode is updated, need to recalculate the checksum */
+                        uint8_t checksum = 0, *p = (uint8_t *) hdr;
+                        hdr->checksum = 0;
+                        FOREACH_ARRAY(c, p, sizeof(UX_CAPSULE_HEADER))
+                                checksum += *c;
+                        hdr->checksum = checksum;
+                }
+
+                if (capsule->Flags == 0 && !efi_guid_equal(&guid, MAKE_GUID_PTR(UX_CAPSULE))) {
+#if defined(__aarch64__) || (defined(__riscv) && __riscv_xlen == 64)
```

Comment:
why check for these archs? please add comment

Reply:
I don't really know, they were there in fwupd since support for those arches was added, with no explanation that I could find. I imagine it's simply because those two architectures don't support those flags.

Reply:
I cargo-cult-ed that in from fwupdate.c -- maybe @vathpela knows?

Reply:
I wonder if he copied it from someone else, who copied it from someone else, who copied it from Kevin Bacon